src/command/archive/get/get.c

```diff
@@ -200,74 +362,149 @@ archiveGetCheck(const StringList *archiveRequestList)
+                }
+                MEM_CONTEXT_END();
 
-        // Loop through the pg history and determine which archiveId to use based on the first file in the list
-        bool found = false;
-        const String *archiveId = NULL;
-        List *cache = NULL;
+                // Loop through pg history and determine which archiveIds to use
+                StringList *archivePathList = NULL;
 
-        for (unsigned int pgIdx = 0; pgIdx < infoPgDataTotal(infoArchivePg(info)); pgIdx++)
-        {
-            InfoPgData pgData = infoPgData(infoArchivePg(info), pgIdx);
+                for (unsigned int pgIdx = 0; pgIdx < infoPgDataTotal(infoArchivePg(info)); pgIdx++)
+                {
+                    InfoPgData pgData = infoPgData(infoArchivePg(info), pgIdx);
 
-            // Only use the archive id if it matches the current cluster
-            if (pgData.systemId == controlInfo.systemId && pgData.version == controlInfo.version)
-            {
-                archiveId = infoPgArchiveId(infoArchivePg(info), pgIdx);
-                cache = lstNewP(sizeof(ArchiveGetFindCache), .comparator = lstComparatorStr);
+                    // Only use the archive id if it matches the current cluster
+                    if (pgData.systemId == controlInfo.systemId && pgData.version == controlInfo.version)
+                    {
+                        const String *archiveId = infoPgArchiveId(infoArchivePg(info), pgIdx);
+                        bool found = true;
 
-                found = archiveGetFind(
-                    strLstGet(archiveRequestList, 0), archiveId, &result, cache, strLstSize(archiveRequestList) == 1);
+                        // If the archiveId is in the past make sure it has some files
+                        if (pgIdx != 0)
+                        {
+                            // Get list of files in the archive path
+                            if (archivePathList == NULL)
+                                archivePathList = storageListP(storageRepoIdx(repoIdx), STORAGE_REPO_ARCHIVE_STR);
 
-                // If the file was found then use this archiveId for the rest of the files
-                if (found)
-                    break;
+                            if (!strLstExists(archivePathList, archiveId))
+                                found = false;
+                        }
+
+                        // If the archiveId is most recent or has files then add it
+                        if (found)
+                        {
+                            ArchiveGetFindCacheArchive cacheArchive =
+                            {
+                                .pathList = lstNewP(sizeof(ArchiveGetFindCachePath), .comparator = lstComparatorStr),
+                            };
+
+                            MEM_CONTEXT_BEGIN(lstMemContext(result.archiveFileMapList))
+                            {
+                                cacheArchive.archiveId = strDup(archiveId);
```

Comment:
This will need to be added to each file in the list for that repo, so rather than dup it each time it makes sense to just assign it to the list context.

Reply:
As discussed, decided not to change this, but added a clarifying comment in d4a09984fb4981.